packages/pwa-kit-runtime/src/ssr/server/express.js

```diff
@@ -461,11 +461,23 @@ export const respondFromBundle = ({req, res, path, redirect = 301}) => {
 }
 
 export const getRuntime = () => {
-    return isRemote()
+    const runtime = isRemote()
         ? RemoteServerFactory
         : // The dev server is for development only, and should not be deployed to production.
           // To avoid deploying the dev server (and all of its dependencies) to production, it exists
           // as an optional peer dependency to this package. The unusual `require` statement is needed
           // to bypass webpack and ensure that the dev server does not get bundled.
           eval('require').main.require('pwa-kit-dev/ssr/server/build-dev-server').DevServerFactory
+
+    // The runtime is a JavaScript object.
+    // Sometimes the runtime APIs are invoked directly as express middlewares.
+    // In order to make sure the "this" keyword always have the correct context,
+    // we bind every single method to have the context of the object itself
+    for (const property in runtime) {
+        if (runtime[property] instanceof Function) {
+            runtime[property] = runtime[property].bind(runtime)
+        }
+    }
```

Comment:
This feels weird to me.
1. Can the methods be rewritten to avoid referencing `this` entirely?
2. If not, can they make use of arrow functions to automatically have the proper `this` context?
3. If not, why is this binding done here? The value of `runtime` is a static object. It seems like it would make more sense to do the binding once, when the object is created, rather than doing it here every time `getRuntime` is called. Doing it every time `getRuntime` is called _also_ has the negative consequence of overwriting properties of an object that we already returned to a previous invocation. If somebody calls it once, modifies the returned value, and then calls it again, then we're potentially resetting their changes in ways they wouldn't expect.

Reply:
Great questions.

> Can the methods be rewritten to avoid referencing this entirely?

Yes, we could refer to the factory Objects themselves, i.e. `DevServerFactory` However that would make the inheritance model harder to extend. I believe ideally we'd like to use `this` so that in the future we can build more features and use the OOP principles to extend the server. I personally would leaning towards using `this`, otherwise I feel like we might be regret some day in the future that we made a choice to limit ourselves to not use `this`...

Reply:
> can they make use of arrow functions to automatically have the proper this context?

I had a similar thought when i first saw this problem, gave it a try to use arrow function, but it doesn't look like that's what we want.

```js
const test = {
  prop: 42,
  func: () => {
    return this.prop;
  },
};

console.log(test.func());
// undefined
```

Arrow function do not have `this` binding (from MDN [doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#arrow_functions_used_as_methods)), so it's the opposite of what we want.

Reply:
> If not, why is this binding done here? The value of runtime is a static object. It seems like it would make more sense to do the binding once, when the object is created, rather than doing it here every time getRuntime is called. Doing it every time getRuntime is called also has the negative consequence of overwriting properties of an object that we already returned to a previous invocation. If somebody calls it once, modifies the returned value, and then calls it again, then we're potentially resetting their changes in ways they wouldn't expect.

Initially this PR added the bindings to when the server factories (DevServerFactory and RemoteServerFactory) are created. See [commit](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/598/commits/87bc5ef1abf9d5b0a3fad9d70896a5068681727b) and [commit](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/598/commits/c7949c7b076957a1c6ed2c9758a89d3c7a037128).

However, Vincent noticed an issue with that. The same function cannot be bound twice (TIL), see https://stackoverflow.com/questions/31656593/javascript-function-bind-override-how-to-bind-it-to-another-object. So we've decided to move the binding to `getRuntime`. Although I haven't put too much thought into this, there might be a better place for this, and it will be a non-blocking change, so this is something we could change after the release.

Ideally, if we used classical inheritance, we would be able to bind methods in the `constructors`, but we can't do that because the way inheritance is setup today, we use objects not classes. And the inheritance is done via `Object.assign`.

I believe there is potential to optimize this, feel free to propose a change! :)